Missing default FHIR package on first run

There seems to be a problem where the following error gets thrown upon the first run of the plugin:

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
    at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
Caused by: java.lang.RuntimeException: There was a problem loading FHIR basic types.
    at au.csiro.redmatch.exporter.HapiReflectionHelper.init(HapiReflectionHelper.java:112)
    at au.csiro.redmatch.exporter.HapiReflectionHelper.<init>(HapiReflectionHelper.java:59)
    at au.csiro.redmatch.lsp.RedmatchLanguageServer.<init>(RedmatchLanguageServer.java:48)
    at au.csiro.redmatch.lsp.Application.main(Application.java:26)
    ... 8 more
Caused by: java.nio.file.NoSuchFileException: /Users/obr187/.fhir/packages/hl7.fhir.r4.core#4.0.1/package
    at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
    at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
    at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
    at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
    at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
    at java.base/java.nio.file.Files.readAttributes(Files.java:1764)
    at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:219)
    at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:276)
    at java.base/java.nio.file.FileTreeWalker.walk(FileTreeWalker.java:322)
    at java.base/java.nio.file.FileTreeIterator.<init>(FileTreeIterator.java:71)
    at java.base/java.nio.file.Files.walk(Files.java:3825)
    at java.base/java.nio.file.Files.walk(Files.java:3879)
    at au.csiro.redmatch.util.FhirUtils.getStructureDefinitions(FhirUtils.java:42)
    at au.csiro.redmatch.exporter.HapiReflectionHelper.init(HapiReflectionHelper.java:68)
    ... 11 more
```

--- a/redmatch-grammar/src/main/java/au/csiro/redmatch/terminology/TerminologyService.java
+++ b/redmatch-grammar/src/main/java/au/csiro/redmatch/terminology/TerminologyService.java
@@ -54,7 +54,7 @@ public class TerminologyService {
   public boolean ontoIndexCheck(VersionedFhirPackage fhirPackage) {
     return onto.isIndexed(REDMATCH_PREFIX + fhirPackage.getName(), fhirPackage.getVersion());
   }
-  public void checkPackage(VersionedFhirPackage fhirPackage, ProgressReporter progressReporter) {
+  public void checkPackage(VersionedFhirPackage fhirPackage, ProgressReporter progressReporter) throws IOException {
     log.debug("Checking if FHIR package " + fhirPackage + " is installed");
     try {
       Instant start = Instant.now();
@@ -87,7 +87,7 @@ public class TerminologyService {
       log.info("Finished checking in: " + DateUtils.prettyPrintMillis(timeElapsed));
     } catch (IOException e) {
       log.error("There was an IO error processing FHIR package " + fhirPackage, e);
-      Thread.currentThread().interrupt();
+      throw e;
     }
   }
 
@@ -109,8 +109,12 @@ public class TerminologyService {
           Triplet<VersionedFhirPackage, CompletableFuture<Void>, ProgressReporter> item = blockingQueue.take();
           VersionedFhirPackage fhirPackage = item.getValue0();
           ProgressReporter progressReporter = item.getValue2();
-          checkPackage(fhirPackage, progressReporter);
-          item.getValue1().complete(null);
+          try {
+            checkPackage(fhirPackage, progressReporter);
+            item.getValue1().complete(null);
+          } catch (IOException e) {
+            Thread.currentThread().interrupt();
+          }
         }
       } catch (InterruptedException e) {
         log.error("Thread was interrupted.", e);

--- a/redmatch-grammar/src/main/java/au/csiro/redmatch/terminology/TerminologyService.java
+++ b/redmatch-grammar/src/main/java/au/csiro/redmatch/terminology/TerminologyService.java
@@ -39,6 +39,8 @@ public class TerminologyService {
   private static final String REDMATCH_PREFIX = "http://redmatch.";
 
   private final InternalApi onto;
+  private FhirContext ctx;
+  private Gson gson;
 
   /**
    * Used to make sure the system does not attempt to index validation code system simultaneously.
@@ -47,6 +49,47 @@ public class TerminologyService {
     new LinkedBlockingDeque<>(10);
 
   private final ExecutorService executor = Executors.newFixedThreadPool(1);
+
+
+  public boolean ontoIndexCheck(VersionedFhirPackage fhirPackage) {
+    return onto.isIndexed(REDMATCH_PREFIX + fhirPackage.getName(), fhirPackage.getVersion());
+  }
+  public void checkPackage(VersionedFhirPackage fhirPackage, ProgressReporter progressReporter) {
+    log.debug("Checking if FHIR package " + fhirPackage + " is installed");
+    try {
+      Instant start = Instant.now();
+      log.debug("Package is not indexed");
+      RedmatchGrammarCodeSystemGenerator generator = new RedmatchGrammarCodeSystemGenerator(gson, ctx);
+      CodeSystem cs = generator.createCodeSystem(fhirPackage, progressReporter);
+
+      if (progressReporter != null) {
+        progressReporter.reportProgress(Progress.reportStart("Indexing code system for FHIR package "
+          + fhirPackage));
+      }
+      Path targetFolder = onto.indexFhirCodeSystem(cs);
+      if (progressReporter != null) {
+        progressReporter.reportProgress(Progress.reportEnd());
+      }
+  
+      Path targetFile = targetFolder.resolve(fhirPackage.getName() + ".json");
+      try (FileWriter fw = new FileWriter(targetFile.toFile())) {
+        if (progressReporter != null) {
+          progressReporter.reportProgress(Progress.reportStart("Saving generated code system for FHIR package "
+            + fhirPackage));
+        }
+        ctx.newJsonParser().setPrettyPrint(true).encodeResourceToWriter(cs, fw);
+        if (progressReporter != null) {
+          progressReporter.reportProgress(Progress.reportEnd());
+        }
+      }
+      Instant finish = Instant.now();
+      long timeElapsed = Duration.between(start, finish).toMillis();
+      log.info("Finished checking in: " + DateUtils.prettyPrintMillis(timeElapsed));
+    } catch (IOException e) {
+      log.error("There was an IO error processing FHIR package " + fhirPackage, e);
+      Thread.currentThread().interrupt();
+    }
+  }
 
   /**
    * Constructor.
@@ -57,6 +100,8 @@ public class TerminologyService {
   public TerminologyService(FhirContext ctx, Gson gson) {
     log.info("Initialising terminology service");
     onto = new InternalApi(gson, ctx);
+    this.ctx = ctx;
+    this.gson = gson;
 
     Runnable indexingTask = () -> {
       try {
@@ -64,41 +109,8 @@ public class TerminologyService {
           Triplet<VersionedFhirPackage, CompletableFuture<Void>, ProgressReporter> item = blockingQueue.take();
           VersionedFhirPackage fhirPackage = item.getValue0();
           ProgressReporter progressReporter = item.getValue2();
-          log.debug("Checking if FHIR package " + fhirPackage + " is installed");
-          try {
-            Instant start = Instant.now();
-            log.debug("Package is not indexed");
-            RedmatchGrammarCodeSystemGenerator generator = new RedmatchGrammarCodeSystemGenerator(gson, ctx);
-            CodeSystem cs = generator.createCodeSystem(fhirPackage, progressReporter);
-
-            if (progressReporter != null) {
-              progressReporter.reportProgress(Progress.reportStart("Indexing code system for FHIR package "
-                + fhirPackage));
-            }
-            Path targetFolder = onto.indexFhirCodeSystem(cs);
-            if (progressReporter != null) {
-              progressReporter.reportProgress(Progress.reportEnd());
-            }
-
-            Path targetFile = targetFolder.resolve(fhirPackage.getName() + ".json");
-            try (FileWriter fw = new FileWriter(targetFile.toFile())) {
-              if (progressReporter != null) {
-                progressReporter.reportProgress(Progress.reportStart("Saving generated code system for FHIR package "
-                  + fhirPackage));
-              }
-              ctx.newJsonParser().setPrettyPrint(true).encodeResourceToWriter(cs, fw);
-              if (progressReporter != null) {
-                progressReporter.reportProgress(Progress.reportEnd());
-              }
-            }
-            Instant finish = Instant.now();
-            long timeElapsed = Duration.between(start, finish).toMillis();
-            log.info("Finished checking in: " + DateUtils.prettyPrintMillis(timeElapsed));
-            item.getValue1().complete(null);
-          } catch (IOException e) {
-            log.error("There was an IO error processing FHIR package " + item.getValue0(), e);
-            Thread.currentThread().interrupt();
-          }
+          checkPackage(fhirPackage, progressReporter);
+          item.getValue1().complete(null);
         }
       } catch (InterruptedException e) {
         log.error("Thread was interrupted.", e);
@@ -137,7 +149,7 @@ public class TerminologyService {
 
     CompletableFuture<Void> future = new CompletableFuture<>();
 
-    if (onto.isIndexed(REDMATCH_PREFIX + fhirPackage.getName(), fhirPackage.getVersion())) {
+    if (ontoIndexCheck(fhirPackage)) {
       if (progressReporter != null) {
         progressReporter.reportProgress(Progress.reportStart("Adding package " + fhirPackage));
       }

--- a/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
+++ b/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
@@ -9,6 +9,7 @@ import au.csiro.redmatch.compiler.RedmatchCompiler;
 import au.csiro.redmatch.exporter.HapiReflectionHelper;
 import au.csiro.redmatch.model.VersionedFhirPackage;
 import au.csiro.redmatch.terminology.TerminologyService;
+import au.csiro.redmatch.util.ProgressReporter;
 import ca.uhn.fhir.context.FhirContext;
 import com.google.gson.Gson;
 import org.apache.commons.logging.Log;
@@ -33,6 +34,7 @@ public class RedmatchLanguageServer implements LanguageServer, LanguageClientAwa
   private LanguageClient client;
   private final RedmatchApi api;
   private final VersionedFhirPackage defaultFhirPackage;
+  private final ProgressReporter progressReporter;
 
   /**
    * Constructor.
@@ -43,8 +45,12 @@ public class RedmatchLanguageServer implements LanguageServer, LanguageClientAwa
     // TODO: would be good to allow users to set the default FHIR package through configuration options
     defaultFhirPackage = new VersionedFhirPackage("hl7.fhir.r4.core", "4.0.1");
     TerminologyService terminologyService = new TerminologyService(ctx, gson);
-    RedmatchCompiler compiler = new RedmatchCompiler(gson, terminologyService, defaultFhirPackage,
-      new LspProgressReporter(this));
+    progressReporter = new LspProgressReporter(this);
+    RedmatchCompiler compiler = new RedmatchCompiler(gson, terminologyService, defaultFhirPackage, progressReporter);
+    if (!terminologyService.ontoIndexCheck(defaultFhirPackage)) {
+      log.info("defaultFhirPackage not detected by index on initialisation");
+      terminologyService.checkPackage(defaultFhirPackage, null);
+    }
     HapiReflectionHelper reflectionHelper = new HapiReflectionHelper(ctx, defaultFhirPackage, terminologyService);
     api = new RedmatchApi(ctx, gson, compiler, reflectionHelper, terminologyService);
     textDocumentService = new RedmatchTextDocumentService(this, terminologyService);

--- a/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
+++ b/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
@@ -35,7 +35,6 @@ public class RedmatchLanguageServer implements LanguageServer, LanguageClientAwa
   private LanguageClient client;
   private final RedmatchApi api;
   private final VersionedFhirPackage defaultFhirPackage;
-  private final ProgressReporter progressReporter;
 
   /**
    * Constructor.
@@ -46,18 +45,9 @@ public class RedmatchLanguageServer implements LanguageServer, LanguageClientAwa
     // TODO: would be good to allow users to set the default FHIR package through configuration options
     defaultFhirPackage = new VersionedFhirPackage("hl7.fhir.r4.core", "4.0.1");
     TerminologyService terminologyService = new TerminologyService(ctx, gson);
-    progressReporter = new LspProgressReporter(this);
+    ProgressReporter progressReporter = new LspProgressReporter(this);
     RedmatchCompiler compiler = new RedmatchCompiler(gson, terminologyService, defaultFhirPackage, progressReporter);
-    if (!terminologyService.ontoIndexCheck(defaultFhirPackage)) {
-      log.info("defaultFhirPackage not detected by index on initialisation");
-      try {
-        terminologyService.checkPackage(defaultFhirPackage, null);
-      } catch (IOException e) {
-        throw new RuntimeException("Tried and failed to load the Default FHIR package.", e);
-      }
-    }
-    HapiReflectionHelper reflectionHelper = new HapiReflectionHelper(ctx, defaultFhirPackage, terminologyService);
-    api = new RedmatchApi(ctx, gson, compiler, reflectionHelper, terminologyService);
+    api = new RedmatchApi(ctx, gson, compiler, defaultFhirPackage, terminologyService, progressReporter);
     textDocumentService = new RedmatchTextDocumentService(this, terminologyService);
     workspaceService = new RedmatchWorkspaceService(this);
   }

--- a/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
+++ b/redmatch-language-server/src/main/java/au/csiro/redmatch/lsp/RedmatchLanguageServer.java
@@ -19,6 +19,7 @@ import org.eclipse.lsp4j.services.*;
 
 import java.util.*;
 import java.util.concurrent.CompletableFuture;
+import java.io.IOException;
 
 /**
  * The Redmatch language server implementation.
@@ -49,7 +50,11 @@ public class RedmatchLanguageServer implements LanguageServer, LanguageClientAwa
     RedmatchCompiler compiler = new RedmatchCompiler(gson, terminologyService, defaultFhirPackage, progressReporter);
     if (!terminologyService.ontoIndexCheck(defaultFhirPackage)) {
       log.info("defaultFhirPackage not detected by index on initialisation");
-      terminologyService.checkPackage(defaultFhirPackage, null);
+      try {
+        terminologyService.checkPackage(defaultFhirPackage, null);
+      } catch (IOException e) {
+        throw new RuntimeException("Tried and failed to load the Default FHIR package.", e);
+      }
     }
     HapiReflectionHelper reflectionHelper = new HapiReflectionHelper(ctx, defaultFhirPackage, terminologyService);
     api = new RedmatchApi(ctx, gson, compiler, reflectionHelper, terminologyService);

--- a/redmatch-runtime/src/main/java/au/csiro/redmatch/RedmatchApi.java
+++ b/redmatch-runtime/src/main/java/au/csiro/redmatch/RedmatchApi.java
@@ -56,20 +56,25 @@ public class RedmatchApi {
 
   private final RedmatchCompiler compiler;
 
-  private final HapiReflectionHelper reflectionHelper;
+  private HapiReflectionHelper reflectionHelper;
 
   private final TerminologyService terminologyService;
+  
+  private VersionedFhirPackage defaultFhirPackage;
+  
+  private ProgressReporter progressReporter;
 
 
   public static final Range zeroZero = new Range(new Position(0, 0), new Position(0, 0));
 
-  public RedmatchApi(FhirContext ctx, Gson gson, RedmatchCompiler compiler, HapiReflectionHelper reflectionHelper,
-                     TerminologyService terminologyService) {
+  public RedmatchApi(FhirContext ctx, Gson gson, RedmatchCompiler compiler, VersionedFhirPackage defaultFhirPackage,
+                     TerminologyService terminologyService, ProgressReporter progressReporter) {
     this.ctx = ctx;
     this.gson = gson;
     this.compiler = compiler;
-    this.reflectionHelper = reflectionHelper;
+    this.defaultFhirPackage = defaultFhirPackage;
     this.terminologyService = terminologyService;
+    this.progressReporter = progressReporter;
   }
 
   /**
@@ -154,6 +159,19 @@ public class RedmatchApi {
     }
 
     log.info("Transforming into FHIR resources using rules " + name);
+    
+    if (reflectionHelper == null) {
+      if (!terminologyService.ontoIndexCheck(defaultFhirPackage)) {
+        log.info("defaultFhirPackage not detected by index on initialisation");
+        try {
+          terminologyService.checkPackage(defaultFhirPackage, progressReporter);
+        } catch (IOException e) {
+          throw new RuntimeException("Tried and failed to load the Default FHIR package.", e);
+        }
+      }
+      reflectionHelper = new HapiReflectionHelper(ctx, defaultFhirPackage, terminologyService);
+    }
+    
     FhirExporter exp = new FhirExporter(document, rows, reflectionHelper, terminologyService,
       compiler.getDefaultFhirPackage());
     return Pair.with(exp.transform(progressReporter, cancelToken), document.getDiagnostics());

